scipy/optimize/_linprog_util.py

```diff
@@ -257,6 +263,31 @@ def _clean_inputs(c, A_ub=None, b_ub=None, A_eq=None, b_eq=None, bounds=None):
             "numerical values, each representing the right hand side of an "
             "equality constraints (row) in A_eq")
 
+    if x0 is not None:
+        try:
+            try:
+                x0 = np.asarray(x0, dtype=float).copy().squeeze()
+            except BaseException:
+                raise TypeError
+            if x0.size == 1:
+                x0 = x0.reshape((-1))
+            if len(x0) == 0 or len(x0.shape) != 1:
+                raise ValueError(
+                    "Invalid input for linprog: x0 should be a 1D array; it "
+                    "must not have more than one non-singleton dimension")
+            if len(x0) != len(c):
+                raise ValueError(
+                    "Invalid input for linprog: x0 and c should contain the "
+                    "same number of elements")
+            if not(np.isfinite(x0).all()):
+                raise ValueError(
+                    "Invalid input for linprog: x0 must not contain values "
+                    "inf, nan, or None")
+        except TypeError:
```

Comment:
this level of try-except is definitely not needed, you can just put the message here at the place where you first raise `TypeError`

Reply:
Deferred to #9671